### PR TITLE
UPSTREAM: Add Ceph volume plugin

### PR DESCRIPTION
**Pull request by markturansky**
Added Ceph volume plugin

### CODE REVIEW

**Comment by markturansky** [comment]
@rootfs @smarterclayton

**Comment by smarterclayton** [comment]
We don't have any checks for upstream conversions so make sure they're all correct (and deep copies)

**Comment by smarterclayton** [comment]
Approved but @derekwaynecarr do a quick review please

**Comment by markturansky** [comment]
I just manually verified the 2 generated deep copy files (v1 and v1beta3).  The conversions are accounted for.

**Comment by rootfs** [comment]
do I miss those files under `pkg/volume/`, like [cephfs.go](https://github.com/GoogleCloudPlatform/kubernetes/pull/6649/files#diff-021042d70c9aac8994638a8d0827b1e0)

**Comment by markturansky** [comment]
Odd, I had to use -f for Git to add those files.  Thanks for pointing that out.

**Comment by smarterclayton** [comment]
Doesn't compile

**Comment by markturansky** [comment]
rebased, fixed compile issues, and pushed anew.

**Comment by markturansky** [comment]
@rootfs would you mind testing this in OpenShift after it gets merged?  The fact that merging was so ugly means I might have missed something.  I have no means to test Ceph.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	}
+	return plugin.newBuilderInternal(spec, pod.UID, mounter, secret)
+}
+
+func (plugin *cephfsPlugin) newBuilderInternal(spec *volume.Spec, podUID types.UID, mounter mount.Interface, secret string) (volume.Builder, error) {
+	id := spec.VolumeSource.CephFS.User
+	if id == "" {
+		id = "admin"
+	}
+	secrete_file := spec.VolumeSource.CephFS.SecretFile
```
Comment (liggitt) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: typo: `secrete`

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+func (plugin *cephfsPlugin) NewBuilder(spec *volume.Spec, pod *api.Pod, _ volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	secret := ""
+	if spec.VolumeSource.CephFS.SecretRef != nil {
+		kubeClient := plugin.host.GetKubeClient()
+		if kubeClient == nil {
+			return nil, fmt.Errorf("Cannot get kube client")
+		}
+
+		secretName, err := kubeClient.Secrets(pod.Namespace).Get(spec.VolumeSource.CephFS.SecretRef.Name)
```
Comment (liggitt) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: @markturansky fyi, when you add API calls to volume plugins, you need to make sure bootstrap policy grants that permission to the node role: https://github.com/openshift/origin/blob/master/pkg/cmd/server/bootstrappolicy/policy.go#L273

In this case, secret gets were already allowed, but it is good to check and add a comment so we can keep track of why the various permissions are needed.

Comment (markturansky) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: Thanks for the tip.  I will need to remember this part.  It bit me in the rear when PV was first merged with OS and no one could get their claims.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 1 above)
Comment (markturansky) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: fixed the typo and added a comment to the upstream PR for the same.

**Comment by markturansky** [comment]
rebased and pushed anew.  One of the two travis tests failed?

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go [hunk 3]
```diff
@@ -207,6 +207,8 @@ type VolumeSource struct {
 	PersistentVolumeClaimVolumeSource *PersistentVolumeClaimVolumeSource `json:"persistentVolumeClaim,omitempty"`
 	// RBD represents a Rados Block Device mount on the host that shares a pod's lifetime
 	RBD *RBDVolumeSource `json:"rbd"`
+	// CephFS represents a Ceph FS mount on the host that shares a pod's lifetime
+	CephFS *CephFSVolumeSource `json:"cephfs,omitempty" description:"Ceph filesystem that will be mounted on the host machine"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: The internal API shouldn't have description fields.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go [hunk 4]
```diff
@@ -541,6 +545,21 @@ type RBDVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// CephFSVolumeSource represents a Ceph Filesystem Mount that lasts the lifetime of a pod
+type CephFSVolumeSource struct {
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: No description fields in this file.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go [hunk 5]
```diff
@@ -541,6 +545,21 @@ type RBDVolumeSource struct {
...
+	// Required: Monitors is a collection of Ceph monitors
+	Monitors []string `json:"monitors,omitempty" description:"a collection of Ceph monitors"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: If this is required, I don't think it should be `omitempty`

@derekwaynecarr

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		if err != nil {
+			err = fmt.Errorf("Couldn't get secret %v/%v err: %v", pod.Namespace, spec.VolumeSource.CephFS.SecretRef, err)
+			return nil, err
+		}
+		for name, data := range secretName.Data {
+			secret = string(data)
+			glog.V(1).Infof("ceph secret info: %s/%s", name, secret)
```
Comment (derekwaynecarr) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: Is this debugging?  Why do we log secret data to a log?

**Comment by derekwaynecarr** [comment]
General question on logging of the secret value seemed wrong...

I am not sure how I feel about merging something untested ;-)

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go [hunk 7]
```diff
@@ -254,10 +256,12 @@ type PersistentVolumeSource struct {
 	// NFS represents an NFS mount on the host
 	NFS *NFSVolumeSource `json:"nfs,omitempty" description:"NFS volume resource provisioned by an admin"`
 	// RBD represents a Rados Block Device mount on the host that shares a pod's lifetime
-	RBD *RBDVolumeSource `json:"rbd" description:"rados block volume that will be mounted on the host machine"`
+	RBD *RBDVolumeSource `json:"rbd,omitempty" description:"rados block volume that will be mounted on the host machine"`
 	// ISCSI represents an ISCSI Disk resource that is attached to a
 	// kubelet's host machine and then exposed to the pod.
-	ISCSI *ISCSIVolumeSource `json:"iscsi" description:"an iSCSI disk resource provisioned by an admin"`
+	ISCSI *ISCSIVolumeSource `json:"iscsi,omitempty" description:"an iSCSI disk resource provisioned by an admin"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go]: You add `omitempty` here but do not add it in the internal API.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go [hunk 8]
```diff
@@ -233,6 +235,8 @@ type PersistentVolumeSource struct {
 	// ISCSIVolumeSource represents an ISCSI resource that is attached to a
 	// kubelet's host machine and then exposed to the pod.
 	ISCSI *ISCSIVolumeSource `json:"iscsi"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: Elsewhere in this PR, `omitempty` is added to the json tag for ISCSI, but not here.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (markturansky) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: Going point.  I will comment on this same line upstream.

Upstream PR is https://github.com/GoogleCloudPlatform/kubernetes/pull/6649

@roots

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go [hunk 9]
```diff
@@ -254,10 +256,12 @@ type PersistentVolumeSource struct {
 	// NFS represents an NFS mount on the host
 	NFS *NFSVolumeSource `json:"nfs,omitempty" description:"NFS volume resource provisioned by an admin"`
 	// RBD represents a Rados Block Device mount on the host that shares a pod's lifetime
-	RBD *RBDVolumeSource `json:"rbd" description:"rados block volume that will be mounted on the host machine"`
+	RBD *RBDVolumeSource `json:"rbd,omitempty" description:"rados block volume that will be mounted on the host machine"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go]: You add `omitempty` here but do not add it elsewhere in this PR.  Maybe you should split out an issue to fix the omitempty tags into a separate PR and make this one depend on that.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go [hunk 10]
```diff
@@ -504,6 +505,14 @@ func validateRBD(rbd *api.RBDVolumeSource) errs.ValidationErrorList {
 	return allErrs
 }
 
+func validateCephFS(cephfs *api.CephFSVolumeSource) errs.ValidationErrorList {
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go]: probably should be named `validateCephFSVolumeSource` for consistency.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go [hunk 11]
```diff
@@ -553,6 +562,14 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) errs.ValidationErrorList
 		numVolumes++
 		allErrs = append(allErrs, validateRBD(pv.Spec.RBD).Prefix("rbd")...)
 	}
+	if pv.Spec.CephFS != nil {
+		numVolumes++
+		allErrs = append(allErrs, validateCephFS(pv.Spec.CephFS).Prefix("cephfs")...)
+	}
+	if pv.Spec.ISCSI != nil {
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go]: This looks to me like a duplicate of the code immediately following this `if` statement.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+	return spec.VolumeSource.CephFS != nil || spec.PersistentVolumeSource.CephFS != nil
+}
+
+func (plugin *cephfsPlugin) GetAccessModes() []api.PersistentVolumeAccessMode {
+	return []api.PersistentVolumeAccessMode{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+		api.ReadWriteMany,
+	}
+
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: spurious newline

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 2 above)
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: @markturansky Are you going to make the change suggested above in another PR?

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: This should probably be at loglevel 3 or 4.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs_test.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	plugMgr := volume.VolumePluginMgr{}
+	plugMgr.InitPlugins(ProbeVolumePlugins(), volume.NewFakeVolumeHost("fake", nil, nil))
+	plug, err := plugMgr.FindPluginByName("kubernetes.io/cephfs")
+	if err != nil {
+		t.Errorf("Can't find the plugin by name")
+	}
+	if plug.Name() != "kubernetes.io/cephfs" {
+		t.Errorf("Wrong name: %s", plug.Name())
+	}
+	if plug.CanSupport(&volume.Spec{Name: "foo", VolumeSource: api.VolumeSource{}}) {
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs_test.go]: Shouldn't there be positive tests for pods that have a normal volume and persistent volume for this plugin?

**Comment by pmorie** [comment]
The `omitempty` tags on PersistentVolumeSource are inconsistent in this PR.  I am leaning towards saying that those ought to be fixed in their own PR that this depends on; it will be much easier to make sure they are right that way.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: @derekwaynecarr Agree, we should not be logging the content of secrets.

**Comment by pmorie** [comment]
@rootfs

**Comment by pmorie** [comment]
I would also really prefer it if the upstream PR had an E2E that we could run against an origin instance and verify this feature that way.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go [hunk 14]
```diff
@@ -444,6 +448,21 @@ type RBDVolumeSource struct {
...
 }
 
+// CephFSVolumeSource represents a Ceph Filesystem Mount that lasts the lifetime of a pod
+type CephFSVolumeSource struct {
+	// Required: Monitors is a collection of Ceph monitors
+	Monitors []string `json:"monitors,omitempty" description:"a collection of Ceph monitors"`
+	// Optional: User is the rados user name, default is admin
+	User string `json:"user,omitempty" description:"rados user name; default is admin; optional"`
+	// Optional: SecretFile is the path to key ring for User, default is /etc/ceph/user.secret
+	SecretFile string `json:"secretFile,omitempty" description:"secretFile is the path to secret for rados user; default is /etc/ceph/user.secret; optional"`
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go]: Nit: descriptions should not be phrased with the name of the field.  Instead, say something like 'path to the secret'

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 6 above)
Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: fixed in upstream

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go (hunk 10 above)
Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go]: seems there is no consistent pattern here yet :)

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go (hunk 14 above)
Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/v1/types.go]: fixed in upstream

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs_test.go (hunk 13 above)
Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs_test.go]: add in upstream

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 12 above)
Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: fixed in upstream

**Comment by rootfs** [comment]
@pmorie @derekwaynecarr I'll get e2e added.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go (hunk 11 above)
Comment (markturansky) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/validation/validation.go]: This was a poor merge on my part and not found upstream.

**Comment by markturansky** [comment]
@rootfs I'll make a new patch from your upstream PR and apply the changes to this after you're done.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go (hunk 3 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: correct, internal apis omit description tags.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go (hunk 5 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: agreed

Comment (rootfs) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api/types.go]: updated in upstream

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go (hunk 2 above)
Comment (markturansky) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/volume/cephfs/cephfs.go]: @pmorie Which changes do you refer to?

**Comment by markturansky** [comment]
This is ready for Round 2 of review.  I redid the patch from the upstream PR to get Huamin's changes.

@rootfs first review, please?

**Comment by rootfs** [comment]
@markturansky LGTM, thanks!

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[merge]
